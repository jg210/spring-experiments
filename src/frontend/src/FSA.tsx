import axios, { CancelTokenSource, AxiosRequestConfig, AxiosResponse } from 'axios';

import _ from 'lodash';

const RATINGS_URL = "http://api.ratings.food.gov.uk";

export interface RatingPercentage {
    rating: string,
    percentage: number
}

export interface LocalAuthority {
    name: string,
    localAuthorityId: number
}

// http://api.ratings.food.gov.uk/help
function fetchFromAPI(url: string, cancelTokenSource: CancelTokenSource | null = null): Promise<AxiosResponse<any>> {
    const config: AxiosRequestConfig = {
        headers: {
            'Accept': 'application/json',
            'x-api-version': 2
        }
    };
    if (!(cancelTokenSource === null)) {
        config.cancelToken = cancelTokenSource.token;
    }
    return axios.get(url, config).then(response => response.data);
}

// http://api.ratings.food.gov.uk/Help/Api/GET-Authorities-basic
export function fetchLocalAuthoritiesJson(): Promise<AxiosResponse<any>> {
    return fetchFromAPI(`${RATINGS_URL}/Authorities/basic`);
}

// http://api.ratings.food.gov.uk/Help/Api/GET-Establishments_name_address_longitude_latitude_maxDistanceLimit_businessTypeId_schemeTypeKey_ratingKey_ratingOperatorKey_localAuthorityId_countryId_sortOptionKey_pageNumber_pageSize
export function fetchEstablishmentsJson(
    localAuthorityId: number,
    cancelTokenSource: CancelTokenSource): Promise<AxiosResponse<any>> {
    const url = `${RATINGS_URL}/Establishments?localAuthorityId=${encodeURIComponent(localAuthorityId.toString())}&pageSize=0`;
    return fetchFromAPI(url, cancelTokenSource);
}

export function extractLocalAuthorities(json: any): LocalAuthority[] {
    return json.authorities.map((authority: {Name: string, LocalAuthorityId: number}) => ({
        name: authority.Name,
        localAuthorityId: authority.LocalAuthorityId
    }));
}

export function ratingsPercentages(establishmentsJson: any): RatingPercentage[] {
    const ratingCounts = new Map<string,number>();
    let totalCount = 0;
    establishmentsJson.establishments.forEach((establishment: {RatingValue: string}) => {
        const rating = formatRating(establishment.RatingValue);
        let oldCount = ratingCounts.get(rating);
        if (oldCount === undefined) {
            oldCount = 0;
        }
        ratingCounts.set(rating, oldCount + 1);
        totalCount++;
    });
    const result: RatingPercentage[] = [];
    ratingCounts.forEach((ratingCount: number, rating: string) => {
        result.push({
             rating: rating,
             percentage: 100 * ratingCount / totalCount
        });
    });
    return _.sortBy(result, "rating");
}

// Convert "RatingValue" from Establishments API to human-readable String.
export function formatRating(ratingValue: string): string {
    if (ratingValue === "AwaitingInspection") {
        return "Awaiting Inspection";
    }
    if (ratingValue === "AwaitingPublication") {
        return "Awaiting Publication";
    }
    if (/^[0-9]+$/.test(ratingValue)) {
        return `${ratingValue}-star`;
    }
    return ratingValue;
}