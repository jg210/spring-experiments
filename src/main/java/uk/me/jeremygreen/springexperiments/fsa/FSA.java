package uk.me.jeremygreen.springexperiments.fsa;

import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;
import uk.me.jeremygreen.springexperiments.fsa.api.LocalAuthority;

import java.util.*;

public class FSA {

    private static final String URL = "http://api.ratings.food.gov.uk";

    private static final HttpHeaders HEADERS = createHeaders();

    private static HttpHeaders createHeaders() {
        final HttpHeaders httpHeaders = new HttpHeaders();
        httpHeaders.set("x-api-version", "2");
        httpHeaders.set("Accept", "application/json");
        return HttpHeaders.readOnlyHttpHeaders(httpHeaders);
    }

    private static <T> T fetchFromApi(
            final String url,
            final Class<T> responseClass) {
        return fetchFromApi(url, responseClass, Collections.emptyMap());
    }

    private static <T> T fetchFromApi(
            final String url,
            final Class<T> responseClass,
            final Map<String,?> params)  {
        final RestTemplate restTemplate = new RestTemplate();
        final HttpEntity entity = new HttpEntity<T>(HEADERS);
        final ResponseEntity<T> response = restTemplate.exchange(
                url, HttpMethod.GET, entity, responseClass, params);
        return response.getBody();
    }

    // http://api.ratings.food.gov.uk/Help/Api/GET-Authorities-basic
    // http://api.ratings.food.gov.uk/Authorities/basic
    public static List<FsaAuthority> fetchAuthorities() {
        final FsaAuthorities fsaAuthorities = fetchFromApi(
                URL + "/Authorities/basic",
                FsaAuthorities.class);
        return Collections.unmodifiableList(fsaAuthorities.getAuthorities());
    }

    // http://api.ratings.food.gov.uk/Help/Api/GET-Establishments_name_address_longitude_latitude_maxDistanceLimit_businessTypeId_schemeTypeKey_ratingKey_ratingOperatorKey_localAuthorityId_countryId_sortOptionKey_pageNumber_pageSize
    public static FsaEstablishments fetchEstablishments(final long fsaAuthorityId) {
        final Map<String,Object> params = new HashMap<>();
        params.put("localAuthorityId", fsaAuthorityId);
        params.put("pageSize", 0);
        return fetchFromApi(
                URL + "/Establishments?localAuthorityId={localAuthorityId}&pageSize={pageSize}",
                FsaEstablishments.class,
                params);
    }

}
