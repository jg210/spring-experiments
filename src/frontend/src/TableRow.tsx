import { RatingPercentage } from "./FSA";

interface TableRowProps {
    ratingPercentage: RatingPercentage;
}

export const TableRow = ({ ratingPercentage }: TableRowProps) => {
    return (
        <tr>
            <td className="tableCell">{ratingPercentage.rating}</td>
            <td>
                <div className="barGraphContainer">
                    <div className="barGraph" data-testid="barGraph" style={{width: `${ratingPercentage.percentage}%`}}/>
                </div>
                <div className="tableCell">{Math.round(ratingPercentage.percentage)}%</div>
            </td>
        </tr>
    );
};
