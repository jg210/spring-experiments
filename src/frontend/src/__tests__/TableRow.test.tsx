import { escapeRegExp } from "lodash";
import { RatingPercentage } from "../FSA";
import { TableRow } from "../TableRow";
import { render, screen, within } from "@testing-library/react";

const matchWholeString = (string: string) => new RegExp("^" + escapeRegExp(string) + "$");

describe("TableRow component", () => {
    it("renders", () => {
        const ratingPercentage : RatingPercentage = {
            rating: "Supercalifragilisticexpialidocious",
            percentage: 23.123
        };
        const { container } = render(
            <table>
                <tbody>
                    <TableRow ratingPercentage={ratingPercentage}/>
                </tbody>
            </table>
        );
        const cells = screen.getAllByRole("cell");
        expect(cells.length).toBe(2);
        const [ ratingCell, percentageCell ] = cells;

        expect(ratingCell).toHaveTextContent(matchWholeString(ratingPercentage.rating));
        expect(ratingCell).toHaveClass("table-cell");

        // The value is rounded to nearest whole percentage.
        expect(percentageCell).toHaveTextContent(matchWholeString('23%'));
        const barGraphContainer = within(percentageCell).getByTestId("barGraphContainer");
        expect(barGraphContainer).toHaveClass("bar-graph-container");
        const barGraph = within(barGraphContainer).getByTestId("barGraph");
        expect(barGraph).toHaveClass("bar-graph");
        expect(barGraph).toHaveStyle({ width: `${ratingPercentage.percentage}%`});

        expect(container).toMatchSnapshot();
    });
});