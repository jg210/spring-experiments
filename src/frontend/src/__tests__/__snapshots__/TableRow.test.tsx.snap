// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableRow component > renders 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="table-cell"
        >
          Supercalifragilisticexpialidocious
        </td>
        <td>
          <div
            class="bar-graph-container"
            data-testid="barGraphContainer"
          >
            <div
              class="bar-graph"
              data-testid="barGraph"
              style="width: 23.123%;"
            />
          </div>
          <div
            class="table-cell"
          >
            23
            %
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
